python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/utilities/device_operators.cuh>
+
+#include <cuda/atomic>
+
+#include <cooperative_groups.h>
+
+#include <limits>
+#include <type_traits>
+
+template <typename T>
+__device__ bool are_all_nans(cooperative_groups::thread_block const& block,
+                             T const* data,
+                             int64_t size)
+{
+  __shared__ bool result;
+
+  if (block.thread_rank()) { result = true; }
+
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    if (not std::isnan(data[idx])) {
+      result = false;
+      break;
+    }
+  }
+
+  block.sync();
```

Comment:
Is this the right amount of synchronisation? Every thread apart from rank 0 writes to the shared `result` variable. If thread-0 just continues on and finds a non-nan entry, could it be that its write is first.

IOW, why not this?
```suggestion
  __shared__ bool result;

  if (block.thread_rank() == 0) { result = true; }
  block.sync()

  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
    if (not std::isnan(data[idx])) {
      result = false;
      break;
    }
  }

  block.sync();
```

Aside, I think `block.size()` is deprecated in favour of `block.num_threads()` ?

Reply:
Fixed.

Replacing `size()` with `num_threads()` will be future work since the new name was recently introduced in 11.6 and it cannot compile with the current CI building environment. https://godbolt.org/z/W7sMqzjvx

Reply:
@PointKernel We build against CUDA 11.8 as of https://github.com/rapidsai/cudf/pull/12457. It is possible to bump our minimum CUDA version required for building to 11.6 if we want.

Reply:
Sorry I wasn't clear. `num_threads()` requires our minimum compiler version to be 11.6 but we still compile libcudf with 11.5 IIRC.

Reply:
We build with 11.8 as of #12457. Currently it is _possible_ to compile with 11.5 on a local machine, but CI and package releases use 11.8. We could bump the minimum version to 11.6, it just requires an update to the contributing guide in CONTRIBUTING.md.

Reply:
Discussed offline. Bumping the minimum CUDA version will be addressed later.